File: src/test/java/org/b0n541/pmcts/mcts/TestGameState.java
package org.b0n541.pmcts.mcts;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class TestGameState implements GameState {

    private final List<GameMove> moves = new ArrayList<>();
    private final int playerCount;

    public TestGameState(final int playerCount) {
        this.playerCount = playerCount;
    }

    private TestGameState(final int playerCount, final List<GameMove> moves, final GameMove move) {
        this.playerCount = playerCount;
        this.moves.addAll(moves);
        this.moves.add(move);
    }

    @Override
    public List<GameMove> getPossibleMoves() {
        return Arrays.asList(TestGameMove.HEADS, TestGameMove.TAILS);
    }

    @Override
    public GameState addMove(final GameMove move) {
        return new TestGameState(playerCount, moves, move);
    }

    @Override
    public boolean isGameFinished() {
        return moves.size() == 100;
    }

    @Override
    public double getGameResult() {
        return moves.stream()
                .filter(move -> move == TestGameMove.TAILS)
                .count();
    }

    @Override
    public int getPlayerCount() {
        return playerCount;
    }

    @Override
    public int getRootPlayerIndex() {
        return 0;
    }

    @Override
    public int getPlayerIndex() {
        return 0;
    }
}

File: src/test/kotlin/net/b0n541/pmcts/mcts/TestGameState.kt
package net.b0n541.pmcts.mcts;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;

public class TestGameState implements GameState<GameMove> {

    private final List<GameMove> moves = new ArrayList<>();
    private final List<String> players;

    public TestGameState(final List<String> players) {
        this.players = Collections.unmodifiableList(players);
    }

    private TestGameState(final List<String> players, final List<GameMove> moves, final GameMove move) {
        this(players);
        this.moves.addAll(moves);
        this.moves.add(move);
    }

    @Override
    public List<String> getPlayers() {
        return players;
    }

    @Override
    public List<GameMove> getPossibleMoves() {
        return List.of(TestGameMove.HEADS, TestGameMove.TAILS);
    }

    @Override
    public GameState addMove(final GameMove move) {
        return new TestGameState(players, moves, move);
    }

    @Override
    public boolean isGameFinished() {
        return moves.size() == 100;
    }

    @Override
    public Map<String, Double> getGameValues() {
        final double headsGameValue = 1.0 *
                moves.stream()
                        .filter(move -> move == TestGameMove.HEADS)
                        .count();
        return Map.of(TestGameMove.HEADS.toString(), headsGameValue, TestGameMove.TAILS.toString(), 100 - headsGameValue);
    }

    @Override
    public GameMove getLastMove() {
        return moves.get(moves.size() - 1);
    }

    @Override
    public String getNextPlayer() {
        return players.get(0);
    }
}
